### PR TITLE
Enhance inspect_cfg

### PR DESCRIPTION
**Pull request by stuartarchibald**
As title. Adds lots of options for helping with CFG views.

Does this sort of thing...
cfg_highlights
needs some tests.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
Does this sort of thing...
cfg_highlights
needs some tests.

**Comment by stuartarchibald** [comment]
note to self, need to fix match for non-exit-1 raise.

File: numba/core/codegen.py [hunk 1]
```diff
@@ -60,8 +62,369 @@ class _CFG(object):
...
+                                                        clean(mark_line))
+                                        new_lines.append(ln)
+                                    if _interleave.python:
+                                        # TODO:
+                                        # +1 for decorator, this probably needs
+                                        # the same thing doing as for the
+                                        # error messages where the decorator
+                                        # is scanned for, its not always +1!
+                                        lidx = int(line) - (firstlineno + 1)
+                                        source_line = src_code[lidx]
```
Comment (sklam) [inline numba/core/codegen.py:350]: There's a off-by-one error. 
```suggestion
                                        source_line = src_code[lidx + 1]
```

File: numba/core/codegen.py [hunk 2]
```diff
@@ -60,8 +62,369 @@ class _CFG(object):
...
+        # LLVM syntax, matches a LLVM debug marker
+        metadata_marker = re.compile(r'.*!dbg\s+(![0-9]+).*')
+        # LLVM syntax, matches a location entry
+        location_expr = (r'.*!DILocation\(line:\s+([0-9]+),'
+                         '\s+column:\s+([0-9]),.*')
+        location_entry = re.compile(location_expr)
+        # LLVM syntax, matches LLVMs internal debug value calls
+        dbg_value = re.compile(r'.*call void @llvm.dbg.value.*')
+        # LLVM syntax, matches tokens for highlighting
+        nrt_incref = re.compile("NRT_incref")
```
Comment (sklam) [inline numba/core/codegen.py:194]: make the pattern a bit more restrictive
```suggestion
        nrt_incref = re.compile(r"@NRT_incref\b")
```

File: numba/core/codegen.py [hunk 3]
```diff
@@ -60,8 +62,369 @@ class _CFG(object):
...
+        nrt_decref = re.compile("NRT_decref")
```
Comment (sklam) [inline numba/core/codegen.py:195]:
```suggestion
        nrt_decref = re.compile(r"@NRT_decref\b")
```

File: numba/core/codegen.py [hunk 4]
```diff
@@ -60,8 +62,369 @@ class _CFG(object):
...
+        nrt_meminfo = re.compile("NRT_MemInfo")
```
Comment (sklam) [inline numba/core/codegen.py:196]:
```suggestion
        nrt_meminfo = re.compile("@NRT_MemInfo")
```

File: numba/core/codegen.py [hunk 5]
```diff
@@ -60,8 +62,369 @@ class _CFG(object):
...
+                                        # +1 for decorator, this probably needs
+                                        # the same thing doing as for the
+                                        # error messages where the decorator
+                                        # is scanned for, its not always +1!
+                                        lidx = int(line) - (firstlineno + 1)
+                                        source_line = src_code[lidx]
+                                        ln = fmt.format(cs['python'], col_span,
+                                                        clean(source_line))
+                                        new_lines.append(ln)
+                                    cur_line, cur_col = line, col
```
Comment (sklam) [inline numba/core/codegen.py]: Too many levels. suggest to make a function for it.

File: numba/core/codegen.py [hunk 6]
```diff
@@ -60,8 +62,369 @@ class _CFG(object):
...
+        port_match = re.compile('.*{(.*)}.*')
+        # DOT syntax, matches the "port" value from a found "port_match"
+        port_jmp_match = re.compile('.*<(.*)>(.*)')
+        # DOT syntax, matches the DOT object (node), captures the entire node
+        node = re.compile(r'.*Node0x.*=".*', re.DOTALL|re.MULTILINE)
+        # LLVM syntax, matches a LLVM debug marker
+        metadata_marker = re.compile(r'.*!dbg\s+(![0-9]+).*')
+        # LLVM syntax, matches a location entry
+        location_expr = (r'.*!DILocation\(line:\s+([0-9]+),'
+                         '\s+column:\s+([0-9]),.*')
```
Comment (sklam) [inline numba/core/codegen.py:189]:
```suggestion
                         r'\s+column:\s+([0-9]),.*')
```

File: numba/core/codegen.py [hunk 7]
```diff
@@ -60,8 +62,369 @@ class _CFG(object):
...
+
+        # A lot of regex is needed to parse the raw dot output. This output
+        # contains a mix of LLVM IR in the labels, and also DOT markup.
+
+        # DOT syntax, matches a "port" (where the tail of an edge starts)
+        port_match = re.compile('.*{(.*)}.*')
+        # DOT syntax, matches the "port" value from a found "port_match"
+        port_jmp_match = re.compile('.*<(.*)>(.*)')
+        # DOT syntax, matches the DOT object (node), captures the entire node
+        node = re.compile(r'.*Node0x.*=".*', re.DOTALL|re.MULTILINE)
```
Comment (sklam) [inline numba/core/codegen.py:184]: dead variable

File: numba/core/codegen.py [hunk 8]
```diff
@@ -60,8 +62,369 @@ class _CFG(object):
...
+            # is only emitted if there's a change in the marker.
+            cur_line, cur_col = -1, -1
+            label = obj['label']
+            name = obj['name']
+            gvid = obj['_gvid']
+            node_ids[gvid] = name
+            # Label is DOT format, it needs the head and tail removing and then
+            # splitting for walking.
+            label = label[1:-1]
+            lines =  label.split('\\l')
```
Comment (sklam) [inline numba/core/codegen.py:245]:
```suggestion
            lines = label.split('\\l')
```

File: numba/core/codegen.py [hunk 9]
```diff
@@ -60,8 +62,369 @@ class _CFG(object):
...
+                            if ld is not None:
+                                # and the metadata is line info... proceed
+                                assert len(ld.groups()) == 2, ld
+                                line, col = ld.groups()
+                                # only emit a new marker if the line number in
+                                # the metadata is "new".
+                                if line != cur_line or col != cur_col:
+                                    if _interleave.lineinfo:
+                                        mfmt = 'Marker %s, Line %s, column %s'
+                                        mark_line =  mfmt % (marker, line, col)
```
Comment (sklam) [inline numba/core/codegen.py:339]:
```suggestion
                                        mark_line = mfmt % (marker, line, col)
```

File: numba/core/codegen.py [hunk 10]
```diff
@@ -60,8 +62,369 @@ class _CFG(object):
...
+
+            # add in the port line at the end of the block if it was present
+            # (this was built right at the top of the parse)
+            if port_line:
+                new_lines.append('<tr>{}</tr>'.format(port_line))
+
+            # If there was data, create a table, else don't!
+            dat = ''.join(new_lines)
+            if dat:
+                tab = ('<table id="%s" BORDER="1" CELLBORDER="0" CELLPADDING="0" '
```
Comment (sklam) [inline numba/core/codegen.py]: line >80ch

**Review by sklam — COMMENTED** [review]
i have some regex and flake8 nitpicks

**Comment by stuartarchibald** [comment]
> i have some regex and flake8 nitpicks

thanks for picking these up, hopefully they are all sorted in 530e828

**Review by sklam — APPROVED** [review]
Approved for merging. Will need follow up patch to address the proper annotation for raising ret.

**Pull request merged**